packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,194 @@
+  // corresponding vpid; each of these is in turn an array of a watcher object
+  // and the arguments associated with it by `watchPromise`.
+  let watchedPromiseTable;
+
+  // defined promise watcher objects indexed by kindHandle
+  let promiseWatcherByKindTable;
+
+  function preparePromiseWatcherTables() {
+    let watcherTableID = syscall.vatstoreGet('watcherTableID');
+    if (watcherTableID) {
+      promiseWatcherByKindTable = convertSlotToVal(watcherTableID);
+    } else {
+      promiseWatcherByKindTable = makeScalarBigMapStore(
+        'promiseWatcherByKind',
+        { durable: true },
+      );
+      watcherTableID = convertValToSlot(promiseWatcherByKindTable);
+      syscall.vatstoreSet('watcherTableID', watcherTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watcherTableID);
+    }
+
+    let watchedPromiseTableID = syscall.vatstoreGet('watchedPromiseTableID');
+    if (watchedPromiseTableID) {
+      watchedPromiseTable = convertSlotToVal(watchedPromiseTableID);
+    } else {
+      watchedPromiseTable = makeScalarBigMapStore('watchedPromises', {
+        keySchema: M.string(), // key is always a vpid
+        durable: true,
+      });
+      watchedPromiseTableID = convertValToSlot(watchedPromiseTable);
+      syscall.vatstoreSet('watchedPromiseTableID', watchedPromiseTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watchedPromiseTableID);
+    }
+  }
+
+  function pseudoThen(p, vpid) {
+    function settle(value, wasFulfilled) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.delete(vpid);
+      for (const watch of watches) {
+        const [watcher, ...args] = watch;
+        Promise.resolve().then(() => {
+          if (wasFulfilled) {
+            if (watcher.onFulfilled) {
+              watcher.onFulfilled(value, ...args);
+            }
+          } else {
+            if (watcher.onRejected) {
+              watcher.onRejected(value, ...args);
+            } else {
+              throw value; // for host's unhandled rejection handler to catch
+            }
+          }
+        });
+      }
+    }
+
+    p.then(
+      res => settle(res, true),
+      rej => settle(rej, false),
+    );
+  }
+
+  function loadWatchedPromiseTable() {
+    const deadPromisesRaw = syscall.vatstoreGet('deadPromises') || '';
+    syscall.vatstoreDelete('deadPromises');
+    const deadPromises = new Set(deadPromisesRaw.split(','));
+
+    for (const [vpid, watches] of watchedPromiseTable.entries()) {
+      if (deadPromises.has(vpid)) {
+        watchedPromiseTable.delete(vpid);
+        for (const watch of watches) {
+          const [watcher, ...args] = watch;
+          Promise.resolve().then(() => {
+            if (watcher.onRejected) {
+              watcher.onRejected(Error('vat upgraded'), ...args);
+            } else {
+              assert.fail('vat upgraded');
```

Comment:
Let's build a shared `Error('vat upgraded')` at like line 99, and then change this `assert.fail` into `throw err`. Using an `assert` makes it read like this is a "shouldn't happen" case, where actually what we're doing is allowing the unhandled rejection handler to observe the unhandled rejection.

Reply:
> One minor change, plus we need an answer from @erights about whether `vat upgraded` should be an Error or a string. After that we're good to land.

Let's try a string. I bet that non-error rejections are only lightly tested at this point, so it wouldn't surprise me if it revealed some such bugs. Which would be a good thing.

Unless there are surprises, I think I like the string idea. Or at least something that is obviously not an Error.

Reply:
Well OK then.

Reply:
@erights I assume you mean to comment on https://github.com/Agoric/agoric-sdk/pull/5130/files#r855445250 ?

For here, an error instance is appropriate. However, we probably should use a shared frozen error built once outside the loop

Reply:
Two additional thoughts: the original reason I switched it from a string to an Error was that eslint whines if you throw a string.  Obviously this can be suppressed with a disabling comment, bit it seems kinda hinky.  Also, I think I'm unclear what the benefit of using a string in this particular case is supposed to be.  I understand flushing out bugs, but is this the right place to be doing that?

Reply:
String is useless here, I think the comment was about the place you changed from string to error. You change shows that the current reliance on string vs error is a little brittle. I'm opening a separate issue for that.

Reply:
What reliance on string vs. error?  I don't understand what you're referring to here.

Reply:
Please see https://github.com/Agoric/agoric-sdk/pull/5130/files#r855510573

Reply:
Oy.  My preference would be to use Error here (and everywhere else there's a throw, but that would be a separate task).  And file a ticket to get rid of the annoying behavior of marshal spewing to the log any time it serializes an Error, which i hate, hate, hate.

Reply:
@erights I guess I'm asking you to reconsider your call above.

Reply:
Sure. Let's have a conversation where we jointly reconsider. Schedule something?

Reply:
@erights and I discussed this and concluded that signaling vat upgrade with a string is preferred to using Error (even better would be to have an exceptional condition indicator object analogous to Error but which is not Error, but we don't have the time to prioritize that right now).  Whatever entity is attending to the vat upgrade event should not be discriminating on the basis of whether the thing thrown was string or an Error however, but should specifically be looking for the particular string (in this case, `'vat upgraded'`) that indicates the condition of interest.

On that basis I'm going to revise this PR to revert to using a string.

Reply:
Can we please make it a `const` at the top level somewhere, and document this string must not change? Then you can throw the const everywhere, and shouldn't have to disable eslint.